File: ruby/058-proverb/proverb.rb
class Proverb
  attr_reader :input, :qualifier

  def initialize(*input, qualifier: '')
    @input = input
    @qualifier = qualifier
  end

  def to_s
    input.each_cons(2)
         .map { |lacked_item, jeopardised_item| "For want of a #{lacked_item} the #{jeopardised_item} was lost." }
         .push("And all for the want of a #{qualifier} #{input.first}.".squeeze(' '))
         .join("\n")
  end
end

File: go/057-proverb/proverb.go
package proverb

import "fmt"

func Proverb(rhyme []string) []string {
	var proverb []string

	if len(rhyme) == 0 {
		return proverb
	}

	for i := range rhyme {
		if i >= len(rhyme)-1 {
			break
		}

		proverb = append(
			proverb,
			fmt.Sprintf("For want of a %s the %s was lost.", rhyme[i], rhyme[i+1]),
		)
	}

	proverb = append(proverb, fmt.Sprintf("And all for the want of a %s.", rhyme[0]))

	return proverb
}
